File: jpp/lab4/second/srcs/doubleWrapper.cpp
#include "doubleWrapper.hpp"

#include <cmath>

namespace doubleWrapper {

DoubleWrapper::DoubleWrapper(const double value) { this->value = value; }
DoubleWrapper::DoubleWrapper() { this->value = 0; }

DoubleWrapper DoubleWrapper::operator+(const DoubleWrapper dw) const {
  return DoubleWrapper(this->value + dw.value);
}

DoubleWrapper DoubleWrapper::operator-(const DoubleWrapper dw) const {
  return DoubleWrapper(this->value - dw.value);
}

DoubleWrapper DoubleWrapper::operator*(const DoubleWrapper dw) const {
  return DoubleWrapper(this->value * dw.value);
}

DoubleWrapper DoubleWrapper::operator/(const DoubleWrapper dw) const {
  return DoubleWrapper(this->value / dw.value);
}

DoubleWrapper DoubleWrapper::operator^(const DoubleWrapper dw) const {
  return DoubleWrapper(std::pow(this->value, dw.value));
}

bool DoubleWrapper::operator<(const DoubleWrapper dw) const {
  return this->value < dw.value;
}

bool DoubleWrapper::operator>(const DoubleWrapper dw) const {
  return this->value > dw.value;
}

}  // namespace doubleWrapper

File: jpp/lab4/javaMvn/src/main/java/wilk/marcin/DoubleWrapper.java
package wilk.marcin;

public class DoubleWrapper implements PolynomialType<DoubleWrapper> {
  private double value;

  public DoubleWrapper(double value) {
    this.value = value;
  }

  @Override
  public DoubleWrapper neutral() {
    return new DoubleWrapper(0.0);
  }

  @Override
  public DoubleWrapper value() {
    return this;
  }

  @Override
  public DoubleWrapper add(DoubleWrapper p) {
    return new DoubleWrapper(this.value + p.value);
  }

  @Override
  public DoubleWrapper sub(DoubleWrapper p) {
    return new DoubleWrapper(this.value - p.value);
  }

  @Override
  public DoubleWrapper prod(DoubleWrapper p) {
    return new DoubleWrapper(this.value * p.value);
  }

  @Override
  public DoubleWrapper div(DoubleWrapper p) {
    return new DoubleWrapper(this.value / p.value);
  }

  @Override
  public DoubleWrapper pow(int p) {
    return new DoubleWrapper(Math.pow(this.value, p));
  }

  @Override
  public boolean gt(DoubleWrapper other) {
    return this.value > other.value;
  }

  @Override
  public boolean lt(DoubleWrapper other) {
    return this.value < other.value;
  }

  @Override
  public boolean equals(Object obj) {
    if (obj == null || !(obj instanceof DoubleWrapper)) {
      return false;
    }

    return this.value == ((DoubleWrapper) obj).value;
  }

  @Override
  public String toString() {
    return Double.toString(value);
  }
}
